src/bvals/cc/bvals_cc_in_one.hpp

```diff
@@ -0,0 +1,61 @@
+//========================================================================================
+// Parthenon performance portable AMR framework
+// Copyright(C) 2020 The Parthenon collaboration
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef BVALS_CC_BVALS_CC_IN_ONE_HPP_
+#define BVALS_CC_BVALS_CC_IN_ONE_HPP_
+
+#include <memory>
+#include <string>
+
+#include "basic_types.hpp"
+#include "bvals/bvals_interfaces.hpp"
+
+namespace parthenon {
+
+template <typename T>
+class MeshData;
+class IndexRange;
+class NeighborBlock;
+namespace cell_centered_bvars {
+void CalcIndicesSetSame(int ox, int &s, int &e, const IndexRange &bounds);
+void CalcIndicesSetFromCoarser(const int &ox, int &s, int &e, const IndexRange &bounds,
+                               const std::int64_t &lx, const int &cng, bool include_dim);
+void CalcIndicesSetFromFiner(int &si, int &ei, int &sj, int &ej, int &sk, int &ek,
+                             const NeighborBlock &nb, MeshBlock *pmb);
+void CalcIndicesLoadSame(int ox, int &s, int &e, const IndexRange &bounds);
+void CalcIndicesLoadToFiner(int &si, int &ei, int &sj, int &ej, int &sk, int &ek,
+                            const NeighborBlock &nb, MeshBlock *pmb);
+TaskStatus SendBoundaryBuffers(std::shared_ptr<MeshData<Real>> &md);
+TaskStatus ReceiveBoundaryBuffers(std::shared_ptr<MeshData<Real>> &md);
+TaskStatus SetBoundaries(std::shared_ptr<MeshData<Real>> &md);
+
+struct BndInfo {
+  int si = 0;
+  int ei = 0;
+  int sj = 0;
+  int ej = 0;
+  int sk = 0;
+  int ek = 0;
+  parthenon::ParArray1D<Real> buf; // comm buffer
+  parthenon::ParArray4D<Real> var; // data variable (could also be coarse array)
+};
+
+using BufferCache_t = ParArray1D<BndInfo>;
```

Comment:
Overall I like this design here. hOwever, the com and var buffers probably need to be changed. Perhaps they should be the shape of a `VariablePack` or something similar?

Reply:
Hmm, why is that?
Our current design guarantees (/requires) that the comm buffers are 1D and the Variable buffers are 4D.
If we'd change the design to go from 56 independent buffer to one large buffer (with subviews or manual indexing), then this may change.

Reply:
When you go from 1 variable to multiple, don't you need one buffer per variable? Or do they get packed into the single contiguous buffer?

Reply:
Correct, our current implementation uses one buffer per variable (vector), so there's room for improvement.
This is also related to introducing support for >1 variables. Ideally, we'd pack all variables in a single buffer and handle with a single call, but that may turn out to warrant a major restructuring (as the "packing" would now need to be implemented natively in the boundary functions, so we could not handle them independently anymore).

Reply:
Keeping track of this in #427